### PR TITLE
expose events in api

### PR DESCRIPTION
**Pull request by pamapa**
Events part of #239 

I would appreciate if the you could make the deprecation. Each deprecation must come with a
replacement code instruction, which i have not looked up yet...

### Checklist

- [x] This PR makes changes to the public API 
- [x] I have included links for closing relevant issue numbers

### CODE REVIEW

File: src/AuthContext.ts [hunk 1]
```diff
@@ -15,6 +16,7 @@ export interface AuthContextProps extends AuthState {
      * UserManager functions. See [UserManager](https://github.com/authts/oidc-client-ts) for more details.
      */
     readonly settings: UserManagerSettings;
+    get events(): UserManagerEvents;
```
Comment (kherock) [inline src/AuthContext.ts:19]: Why the `get` type? I think `readonly` makes more sense here

**Review by kherock — COMMENTED** [review]
As far as deprecating the props from `AuthProvider` goes, I'm not sure if it is actually a good idea yet. I think it's ok to not do anything about them for now.
My issue with them is that I'm not sure how to make them useful when the provider is typically defined at the root of my application where I usually avoid defining any business logic.

Anyway, the migration from props to event listeners defined inside hooks would look something like this:

* `onSigninCallback`:
```js
useEffect(() => {
  if (hasAuthParams())
    return events.addUserLoaded(() => { ... })
}, [events])
```
* `onRemoveUser`:
```js
useEffect(() => events.addUserUnloaded(() => {...}), [events])
```

The `onSignout*` props seem very odd to me since it seems to encourage discarding the promises returned from the `signout*()` functions in the AuthContext. They also don't resolve until they complete, making them effectively the same as the `onRemoveUser` prop. However, writing effects that run based on the value of `activeNavigator` should work:

* `onSignoutPopup`:
```js
// runs upon activation of the popup navigator
useEffect(() => {
  if (activeNavigator === 'signoutPopup') {
    // do stuff as soon as the popup is open
    // ...
    return () => {
      // do stuff after the popup closes
      // ...
    }
  }
}, [activeNavigator])
```

* `onSignoutRedirect`: I'm not sure why this exists since it shouldn't ever run as it waits for the redirect promise to resolve. You can use the same approach as above but using the `'signoutRedirect'` string instead

**Comment by pamapa** [comment]
One of my use-case is using those functions for window navigation. Before sign-in we can remember the location and after the login we can re-apply it, because that is where the user wanted to go to. As we have multiple applications and they shall all behave the same those functions make it easy. I can prepare the settings in a shared library like:
```javascript
{
  // ...
  onSigninCallback:() => {
    // go to previous location
    const prevPath = window.sessionStorage.getItem("PREV_PATH") || process.env.PUBLIC_URL || "/";
    window.sessionStorage.removeItem("PREV_PATH");
    window.location.replace(prevPath);
  },
  onRemoveUser: () => {
    // go to home location
    const home = process.env.PUBLIC_URL || "/";
    window.location.replace(home);
  }
}
```
For this kind of code using `useEffect` + `events` seems not so handy. Lets **not** deprecate those functions, even if there is a small redundancy. Also when looking at other libraries, they have the same callbacks (e.g. [auth0](https://github.com/auth0/auth0-react/blob/70743cb5a572097508bad931b42b05a8de65af01/src/auth0-provider.tsx#L249)).

For the `onSignout*` props i am not so sure. Maybe we should deprecate them (`onSignoutRedirect` + `onSignoutPopup`)...

**Comment by pamapa** [comment]
Proposal: we only deprecate `onSignoutPopup`.

**Comment by kherock** [comment]
`onSignoutRedirect` is unusable in 2.0's current implementation unless I'm missing something, so I'm ok with deprecating them both.

**Comment by pamapa** [comment]
ready for merging

File: src/AuthProvider.tsx [hunk 2]
```diff
@@ -47,23 +47,24 @@ export interface AuthProviderProps extends UserManagerSettings {
...
      * On remove user hook. Can be a async function.
-     */
-    onRemoveUser?: () => Promise<void> | void;
-
-    /**
-     * On sign out redirect hook. Can be a async function.
      * Here you can change the url after the logout.
+     *
      * ```jsx
      * const onSignOutRedirect = (): void => {
```
Comment (kherock) [inline src/AuthProvider.tsx:53]: This example should be renamed as well just to make it clear that this is a function you'd be able to pass as the `onRemoveUser` prop

**Review by kherock — APPROVED** [review]
Thanks! The only additional thing I'd like to see from this PR is an example in the main README for creating event listeners with `useEffect`. It's not immediately apparent that you can return the "add listener" return value to clean up the hook.

Maybe something like this:

### Adding event listeners

The underlying [`UserManagerEvents`](https://authts.github.io/oidc-client-ts/classes/UserManagerEvents.html) instance can be imperatively managed with the `useAuth` hook.

```jsx
// src/App.jsx
import React from "react";
import { useAuth } from "react-oidc-context";

function App() {
    const auth = useAuth();

    React.useEffect(() => {
      // (the return is important - addAccessTokenExpiring() returns a cleanup function)
      return auth.events.addAccessTokenExpiring(() => {
        if (alert("You're about to be signed out due to inactivity. Press continue to stay signed in.")) {
          auth.signinSilent()
        }
      })
    }, [auth.events])
    
    return <button onClick={auth.signinRedirect}>Log in</button>;
}

export default App;
```

**Pull request merged**